docs/proposals/rescheduling-for-critical-pods.md

```diff
@@ -0,0 +1,116 @@
+
+Rescheduler will ensure that critical addons are always scheduled.
+In the first version it will implement only this policy, but later we may want to introduce other policies.
+It will be a standalone component running on master machine similarly to scheduler.
+Those components will share common logic (initially rescheduler will in fact import some of scheduler packages).
+
+### Guaranteed scheduling of critical addons
+
+Rescheduler will observe critical addons (running in `kube-system` namespace
+with annotation `alpha.kubernetes.io/service-type` set to `CRITICAL`).
+If one of them is marked by scheduler as unschedulable (pod condition `PodScheduled` set to `false`, the reason set to `Unschedulable`)
+the component will try to find a space for the addon by evicting some pods and then the scheduler will schedule the addon.
+
+#### Scoring nodes
+
+Initially we want to choose a random node with enough capacity
+(chosen as described in [Evicting pods](rescheduling-for-critical-pods.md#evicting-pods)) to schedule given addons.
+Later we may want to introduce some heuristic:
+* minimize number of evicted pods with violation of disruption budget or shortened termination grace period
+* minimize number of affected pods by choosing a node on which we have to evict less pods
+* increase probability of scheduling of evicted pods by preferring a set of pods with the smallest total sum of requests
+* avoid nodes which are ‘non-drainable’ (according to drain logic), for example on which there is a pod which doesn’t belong to any RC/RS/Deployment
+
+#### Evicting pods
+
+There are 2 mechanism which possibly can delay a pod eviction: Disruption Budget and Termination Grace Period.
+
+While removing a pod we will try to avoid violating Disruption Budget, though we can’t guarantee it
+since there is a chance that it would block this operation for longer period of time.
+We will also try to respect Termination Grace Period, though without any guarantee.
+In case we have to remove a pod with termination grace period longer than 10s it will be shortened to 10s.
+
+The proposed order while choosing a node to schedule a critical addon and pods to remove:
+1. a node where the critical addon pod can fit after evicting only pods satisfying both
+(1) their disruption budget will not be violated by such eviction and (2) they have grace period <= 10 seconds
+1. a node where the critical addon pod can fit after evicting only pods whose disruption budget will not be violated by such eviction
+1. any node where the critical addon pod can fit after evicting some pods
+
+### Interaction with Scheduler
+
+To avoid situation when Scheduler will schedule another pod into the space prepared for the critical addon,
+the chosen node has to be temporarily excluded from a list of nodes considered by Scheduler while making decisions.
+For this purpose the node will get a temporary taint “CriticalAddonsOnly”
+and each critical addon has to have defined toleration for this taint.
```

Comment:
We could make toleration an updatable field, perhaps, as long as kubelet
doesn't hash on it.

On Jul 24, 2016 12:21 PM, "David Oppenheimer" <notifications@github.com>
wrote:

> In docs/proposals/rescheduling-for-critical-pods.md
> <https://github.com/kubernetes/kubernetes/pull/29195#discussion_r71993754>
> :
>
> > +since there is a chance that it would block this operation for longer period of time.
> > +We will also try to respect Termination Grace Period, though without any guarantee.
> > +In case we have to remove a pod with termination grace period longer than 10s it will be shortened to 10s.
> > +
> > +The proposed order while choosing a node to schedule a critical addon and pods to remove:
> > +1. a node where the critical addon pod can fit after evicting only pods satisfying both
> > +(1) their disruption budget will not be violated by such eviction and (2) they have grace period <= 10 seconds
> > +1. a node where the critical addon pod can fit after evicting only pods whose disruption budget will not be violated by such eviction
> > +1. any node where the critical addon pod can fit after evicting some pods
> > +
> > +### Interaction with Scheduler
> > +
> > +To avoid situation when Scheduler will schedule another pod into the space prepared for the critical addon,
> > +the chosen node has to be temporarily excluded from a list of nodes considered by Scheduler while making decisions.
> > +For this purpose the node will get a temporary taint “CriticalAddonsOnly”
> > +and each critical addon has to have defined toleration for this taint.
>
> Pods have to specify it, but it's essentially just an extra line in the
> PodSpec. That does raise an issue around how we will roll out this
> feature--we need to update all the pod templates for the critical pods
> before we enable this feature. (We can leave the critical pods that are
> already running as-is, because the toleration isn't needed we need to
> schedule a new instance on a node with a taint.)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29195/files/a135c597401e8d9e12e19845600b8061ffdb4b7c#r71993754>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVGrZHF5mBLsoOY65Z2iZcSac7v38ks5qY7segaJpZM4JPqoM>
> .
>